flask_dance/contrib/twitch.py

```diff
@@ -0,0 +1,111 @@
+        login_url (str, optional): The URL path for the ``login`` view. Defaults to None.
+        base_url (str, optional): Base URL for the Twitch API. Defaults to "{Twitch API}/helix/".
+        authorization_url (str, optional): URL for the Twitch API login endpoint. Defaults to "{Twitch API}/oauth2/authorize".
+        token_url (str, optional): URL for Twitch token API. Defaults to "{Twitch API}/oauth2/token".
+        token_url_params (dict, optional): Options to send to the token URL. Defaults to {"include_client_id": True}.
+        authorized_url (str, optional): The URL path for the ``authorized`` view. Defaults to None.
+        session_class (class, optional):  The class to use for creating a Requests session. Defaults to None.
+        storage (class, optional): A token storage class, or an instance of a token storage class, to use for this blueprint. Defaults to None.
+
+    Raises:
+        MissingConfigException: If client_id or client_secret is missing when request is made.
+
+    Returns:
+        :class:`~flask_dance.consumer.OAuth2ConsumerBlueprint` A :ref:`blueprint <flask:blueprints>` to attach to your Flask app.
+    """
+    twitch_bp = OAuth2ConsumerBlueprint(
+        "twitch",
+        __name__,
+        client_id=client_id,
+        client_secret=client_secret,
+        scope=scope,
+        base_url=base_url,
+        authorization_url=authorization_url,
+        token_url=token_url,
+        token_url_params=token_url_params,
+        redirect_url=redirect_url,
+        redirect_to=redirect_to,
+        login_url=login_url,
+        authorized_url=authorized_url,
+        session_class=session_class,
+        storage=storage,
+    )
+    if twitch_bp.client_id is None:
+        twitch_bp.from_config["client_id"] = "TWITCH_OAUTH_CLIENT_ID"
+
+    if twitch_bp.client_secret is None:
+        twitch_bp.from_config["client_secret"] = "TWITCH_OAUTH_CLIENT_SECRET"
+
+    # TODO: The key won't auto renew. See https://github.com/singingwolfboy/flask-dance/issues/35 I think this will work but needs a test.
+    twitch_bp.auto_refresh_url = twitch_bp.token_url
+    twitch_bp.auto_refresh_kwargs = {
+        "client_id": twitch_bp.client_id,
+        "client_secret": twitch_bp.client_secret,
+    }
+
+    @twitch_bp.before_app_request
+    def set_applocal_session():
+
+        ctx = stack.top
+
+        if twitch_bp.client_id is None:
+            raise MissingConfigException(
+                f"client_id is required, currently is {twitch_bp.client_id}. Type: {type(twitch_bp.client_id)}"
+            )
+
+        if twitch_bp.client_secret is None:
+            raise MissingConfigException(
+                f"client_secret is required, currently is  is {twitch_bp.client_secret}. Type: {type(twitch_bp.client_secret)}"
+            )
+
+        twitch_bp.session.headers.update({"client-id": twitch_bp.client_id})
```

Comment:
Ah, so Twitch wants the OAuth client ID to appear in every API call? Interesting... and not compliant with the spec, as far as I know, but that's not unusual. Out of curiousity, can you link me to some documentation for their API that requests this?

Reply:
You can see it in the very first example here: https://dev.twitch.tv/docs/api/

I've not done an exhaustive search on every call but every one that I've looked at or have used requires it: 
https://dev.twitch.tv/docs/api/reference#start-commercial
https://dev.twitch.tv/docs/api/reference#search-channels
https://dev.twitch.tv/docs/api/reference#check-user-subscription

Reply:
Here's an announcement about this: https://blog.twitch.tv/en/2016/05/05/client-id-required-for-kraken-api-calls-afbb8e95f843/

Reply:
I can see why they would do that on the back end. Using the client-id as a primary key would make data-store lookup way faster. I'm sure they have a _few users_ out there. I'll mark this as resolved for now. Thanks!